transport/socks5/packet_dialer.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package socks5
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"golang.org/x/net/proxy"
+)
+
+type packetDialerAdaptor struct {
+	endpoint transport.PacketEndpoint
+}
+
+func (d *packetDialerAdaptor) Dial(network, addr string) (c net.Conn, err error) {
+	return d.endpoint.Connect(context.Background())
+}
+
+func (d *packetDialerAdaptor) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	return d.endpoint.Connect(ctx)
+}
+
+var _ proxy.Dialer = (*packetDialerAdaptor)(nil)
+var _ proxy.ContextDialer = (*packetDialerAdaptor)(nil)
+
+// NewPacketDialer creates a client that routes connections to a SOCKS5 proxy listening at
+// the given [transport.PacketEndpoint].
+func NewPacketDialer(endpoint transport.PacketEndpoint) (*PacketDialer, error) {
```

Comment:
It seems the app might use `PacketDialer` more often. Maybe we can get rid of `PacketListener`?

Reply:
We can't get rid of `PacketListener` because we use it. We need to think of the implications of using a `PacketDialer` instead. For instance, it will require additional bookkeeping in the client to send packets from the same source/UDP conn to different bound net.Conns. That consumes more memory and file descriptors, which I don't know if it will be an issue or not.